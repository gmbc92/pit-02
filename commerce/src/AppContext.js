import { createContext } from "react";

const AppContext = createContext();

const initialState = {
  loggedUser: null,
  products: [],
  cart: [],
  wishlist: [],
  search: "",
};

const Types = {
  SET_LOGGED_USER: "SET_LOGGED_USER",
  SET_PRODUCTS: "SET_PRODUCTS",
  SET_CART: "SET_CART",
  SET_WISHLIST: "SET_WISHLIST",
  SET_SEARCH: "SET_SEARCH",
  TOGGLE_WISHLIST: "TOGGLE_WISHLIST",
};

const AppReducer = (state, action) => {
  switch (action.type) {
    case Types.SET_LOGGED_USER: {
      return {
        ...state,
        loggedUser: action.payload,
      };
    }

    case Types.SET_PRODUCTS: {
      return {
        ...state,
        products: action.payload,
      };
    }

    case Types.SET_CART: {
      return {
        ...state,
        cart: action.payload,
      };
    }

    case Types.SET_SEARCH: {
      return {
        ...state,
        search: action.payload,
      };
    }

    case Types.SET_WISHLIST: {
      return {
        ...state,
        wishlist: action.payload,
      };
    }

    case Types.TOGGLE_WISHLIST: {
      const product = action.payload;
      const alreadyOnWishList = state.wishlist.find(
        (wish) => wish._id === product._id
      );
      let wishlist = [...state.wishlist];

      if (alreadyOnWishList) {
        wishlist = wishlist.filter((wish) => wish._id !== product._id);
      } else {
        wishlist = [...wishlist, product];
      }

      return {
        ...state,
        wishlist,
      };
    }

    default: {
      return state;
    }
  }
};

export { initialState, Types, AppReducer };

export default AppContext;
