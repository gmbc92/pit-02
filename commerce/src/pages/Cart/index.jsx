import React, { useContext, useState } from "react";
import ClayLayout from "@clayui/layout";
import ClayButton, { ClayButtonWithIcon } from "@clayui/button";
import { useModal } from "@clayui/modal";

import Modal from "../../components/modal";
import AppContext, { Types } from "../../AppContext";
import Page from "../../components/page";
import ProductCartInfo from "../../components/Product/ProductCartInfo";
import axios from "../../utils/axios";

const formatedMoney = (value, currencyCode = "USD") => {
  return value.toLocaleString("en-US", {
    style: "currency",
    currency: currencyCode,
  });
};

const CartProduct = ({
  cart: { product, productId, ...cart },
  onUpdateCart,
}) => {
  const [{ cart: cartProducts }, dispatch] = useContext(AppContext);

  const onRemoveCartItem = async () => {
    await axios.delete(`/cart/${productId}`);

    const updateCartProducts = cartProducts.filter(
      (cartProduct) => cartProduct.productId !== productId
    );

    dispatch({
      type: Types.SET_CART,
      payload: updateCartProducts,
    });
  };

  return (
    <ClayLayout.Row className="cart-product-info">
      <ClayLayout.Col>
        <img
          draggable="false"
          alt="product x"
          width={100}
          height={100}
          src={product.images[0].url}
        />
      </ClayLayout.Col>
      <ClayLayout.Col>
        <h2>{product.name}</h2>
        <ClayButton
          displayType="secondary"
          onClick={() => onUpdateCart(cart, product)}
        >
          Update Cart
        </ClayButton>
      </ClayLayout.Col>
      <ClayLayout.Col>
        <h3 className='product-value'>
          {`${formatedMoney(
            product.price.value * cart.quantity,
            product.price.currencyCode
          )}`}
        </h3>
        <ClayButtonWithIcon
          onClick={onRemoveCartItem}
          symbol="trash"
          displayType="secondary"
        />
      </ClayLayout.Col>
    </ClayLayout.Row>
  );
};

const Cart = () => {
  const [visible, setVisible] = useState(false);
  const [productState, setProductState] = useState({});
  const [product, setProduct] = useState();

  const { observer, onClose } = useModal({
    onClose: () => setVisible(false),
  });

  const [{ cart }, dispatch] = useContext(AppContext);

  const onSubmit = async () => {
    const cartProduct = { ...productState };

    delete cartProduct._id;

    const response = await axios.post("/cart", {
      ...cartProduct,
      productId: product._id,
    });

    dispatch({
      type: Types.SET_CART,
      payload: response.data,
    });

    onClose();
  };

  const onUpdateCart = (cart, product) => {
    setVisible(true);
    setProduct(product);
    setProductState(cart);
  };

  const getTotalValue = () => {
    let totalValue = 0;
    let currencyCode = "USD";

    if (cart.length) {
      cart.forEach((item) => {
        totalValue = totalValue + item.product.price.value * item.quantity;
        currencyCode = item.product.price.currencyCode;
      });
    }

    return [totalValue, currencyCode];
  };

  const [totalValue, currencyCode] = getTotalValue();

  return (
    <Page title="My Cart">
      <hr />

      {cart.map((cartProduct, index) => (
        <div key={index} className="mb-2">
          <CartProduct onUpdateCart={onUpdateCart} cart={cartProduct} />
          <hr />
        </div>
      ))}

      <h3>Order Summary</h3>

      <ClayLayout.Row>
        <ClayLayout.Col size={8}>
          <h3>Total</h3>
        </ClayLayout.Col>
        <ClayLayout.Col>
          <h3 className="total-product-value">
            {formatedMoney(totalValue, currencyCode)}
          </h3>
        </ClayLayout.Col>
      </ClayLayout.Row>
      <ClayLayout.Row>
        <ClayLayout.Col size={8}>
          <h3>Tax</h3>
        </ClayLayout.Col>
        <ClayLayout.Col>
          <h3>0</h3>
        </ClayLayout.Col>
      </ClayLayout.Row>

      {cart.length > 0 && (
        <ClayLayout.Row className="mt-4">
          <ClayLayout.Col>
            <ClayButton className="place-order">Place Order</ClayButton>
          </ClayLayout.Col>
        </ClayLayout.Row>
      )}

      <Modal
        observer={observer}
        visible={visible}
        onClose={onClose}
        submitText="Save"
        title="Product Cart"
        onSubmit={onSubmit}
      >
        <ProductCartInfo
          product={product}
          productState={productState}
          setProductState={setProductState}
        />
      </Modal>
    </Page>
  );
};

export default Cart;
