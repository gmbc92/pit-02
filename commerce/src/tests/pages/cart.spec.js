import Cart from "../../pages/Cart";
import AppContext, { initialState } from "../../AppContext";
import React from "react";

import { render } from "@testing-library/react";

const cartArray = [
  {
    _id: "605f45d1c19bce0e2c2ac892",
    productId: "6055f8fabedbed186f70ca03",
    size: "",
    color: "",
    quantity: 1,
    product: {
      _id: "6055f8fabedbed186f70ca03",
      prices: {
        price: {
          value: 80,
          currencyCode: "USD",
        },
      },
      options: [],
      name: "Black Hat",
      path: "/next-js-enamel-mug/",
      images: [
        {
          isDefault: true,
          _id: "6055f8fabedbed186f70ca04",
          url:
            "https://cdn11.bigcommerce.com/s-qfzerv205w/images/stencil/original/products/115/489/Hat-front-black__72990.1603748583.png",
          alt: "",
        },
      ],
      id: 115,
      slug: "next-js-enamel-mug",
      price: {
        value: 80,
        currencyCode: "USD",
      },
    },
  },
];

const CartWrapper = ({
  children,
  state = initialState,
  dispatch = jest.fn(),
}) => (
  <AppContext.Provider value={[state, dispatch]}>
    {children}
  </AppContext.Provider>
);

describe("Cart", () => {
  it("renders cart empty", () => {
    const { container } = render(
      <CartWrapper>
        <Cart />
      </CartWrapper>
    );

    expect(container.querySelector(".total-product-value").textContent).toBe(
      "$0.00"
    );
    expect(container.querySelector(".place-order")).toBeNull();
  });

  it("renders cart with one product", () => {
    const state = {
      ...initialState,
      cart: cartArray,
    };

    const { container } = render(
      <CartWrapper state={state}>
        <Cart />
      </CartWrapper>
    );

    const cartProductInfos = container.querySelectorAll(".cart-product-info");

    expect(container.querySelector(".place-order")).toBeTruthy();
    expect(cartProductInfos.length).toBe(1);
    expect(
      cartProductInfos[0].querySelector(".product-value").textContent
    ).toBe("$80.00");
    expect(container.querySelector(".total-product-value").textContent).toBe(
      "$80.00"
    );
  });

  it("renders cart with many products", () => {
    const newProduct = {
      _id: "605f4774c19bce0e2c2ac893",
      productId: "6055f8fabedbed186f70ca35",
      size: "M",
      color: "White",
      quantity: 5,
      product: {
        _id: "6055f8fabedbed186f70ca35",
        prices: {
          price: {
            value: 44,
            currencyCode: "USD",
          },
        },
        name: "Unisex Skinny Joggers",
        images: [
          {
            isDefault: false,
            _id: "6055f8fabedbed186f70ca36",
            url:
              "https://cdn11.bigcommerce.com/s-qfzerv205w/images/stencil/original/products/126/423/mockup-9c5e3634__60853.1603747807.png",
            alt: "",
          },
        ],
        id: 126,
        slug: "unisex-skinny-joggers",
        price: {
          value: 44,
          currencyCode: "USD",
        },
        __v: 0,
      },
    };

    const state = {
      ...initialState,
      cart: [...cartArray, newProduct],
    };

    const { container, asFragment } = render(
      <CartWrapper state={state}>
        <Cart />
      </CartWrapper>
    );

    const cartProductInfos = container.querySelectorAll(".cart-product-info");

    expect(container.querySelector(".place-order")).toBeTruthy();
    expect(cartProductInfos.length).toBe(2);
    expect(
      cartProductInfos[0].querySelector(".product-value").textContent
    ).toBe("$80.00");
    expect(
      cartProductInfos[1].querySelector(".product-value").textContent
    ).toBe("$220.00");

    expect(container.querySelector(".total-product-value").textContent).toBe(
      "$300.00"
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
