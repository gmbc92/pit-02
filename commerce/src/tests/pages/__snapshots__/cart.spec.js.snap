// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart renders cart with many products 1`] = `
<DocumentFragment>
  <div
    class="mt-4 container"
  >
    <h1>
      My Cart
    </h1>
    <section>
      <hr />
      <div
        class="mb-2"
      >
        <div
          class="row cart-product-info"
        >
          <div
            class="col"
          >
            <img
              alt="product x"
              draggable="false"
              height="100"
              src="https://cdn11.bigcommerce.com/s-qfzerv205w/images/stencil/original/products/115/489/Hat-front-black__72990.1603748583.png"
              width="100"
            />
          </div>
          <div
            class="col"
          >
            <h2>
              Black Hat
            </h2>
            <button
              class="btn btn-secondary"
              type="button"
            >
              Update Cart
            </button>
          </div>
          <div
            class="col"
          >
            <h3
              class="product-value"
            >
              $80.00
            </h3>
            <button
              class="btn btn-monospaced btn-secondary"
              type="button"
            >
              <svg
                class="lexicon-icon lexicon-icon-trash"
                role="presentation"
              >
                <use
                  xlink:href="#trash"
                />
              </svg>
            </button>
          </div>
        </div>
        <hr />
      </div>
      <div
        class="mb-2"
      >
        <div
          class="row cart-product-info"
        >
          <div
            class="col"
          >
            <img
              alt="product x"
              draggable="false"
              height="100"
              src="https://cdn11.bigcommerce.com/s-qfzerv205w/images/stencil/original/products/126/423/mockup-9c5e3634__60853.1603747807.png"
              width="100"
            />
          </div>
          <div
            class="col"
          >
            <h2>
              Unisex Skinny Joggers
            </h2>
            <button
              class="btn btn-secondary"
              type="button"
            >
              Update Cart
            </button>
          </div>
          <div
            class="col"
          >
            <h3
              class="product-value"
            >
              $220.00
            </h3>
            <button
              class="btn btn-monospaced btn-secondary"
              type="button"
            >
              <svg
                class="lexicon-icon lexicon-icon-trash"
                role="presentation"
              >
                <use
                  xlink:href="#trash"
                />
              </svg>
            </button>
          </div>
        </div>
        <hr />
      </div>
      <h3>
        Order Summary
      </h3>
      <div
        class="row"
      >
        <div
          class="col-8"
        >
          <h3>
            Total
          </h3>
        </div>
        <div
          class="col"
        >
          <h3
            class="total-product-value"
          >
            $300.00
          </h3>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-8"
        >
          <h3>
            Tax
          </h3>
        </div>
        <div
          class="col"
        >
          <h3>
            0
          </h3>
        </div>
      </div>
      <div
        class="row mt-4"
      >
        <div
          class="col"
        >
          <button
            class="place-order btn btn-primary"
            type="button"
          >
            Place Order
          </button>
        </div>
      </div>
    </section>
  </div>
</DocumentFragment>
`;
